### ISSUE TITLE
Question about copy and how it affects string output

### ISSUE THREAD

**Issue by tapir** [ericlagergren/decimal#59]
Below is a small example and it's output

```go
package main

import (
	"fmt"

	"github.com/ericlagergren/decimal"
)

const (
	precision = 24
	scale     = 12
)

var defaultCxt = decimal.Context{
	OperatingMode: decimal.Go,
	Precision:     precision,
	RoundingMode:  decimal.ToNearestEven,
	Traps:         ^(decimal.Inexact | decimal.Rounded | decimal.Subnormal),
}

func main() {
	tmp := decimal.WithContext(defaultCxt)
	tmp.SetString("0.0")
	fmt.Println(new(decimal.Big).Copy(tmp).Quantize(scale).String())
	fmt.Println(tmp.Quantize(scale).String())
}
```

```
0E-12
0
```

I was wondering why I'm seeing a scientific notation when I do a copy?

Thanks

**Comment by ericlagergren** [comment]
When converting a decimal to a string, the GDA spec says to convert the decimal to scientific notation _unless_ the exponent (i.e., the negated scale) is <= 0 _and_ the adjusted exponent (i.e., the exponent plus the length of the coefficient minus one) is >= -6. This is the default behavior for the `String` method, assuming the operating mode is `GDA` (the default).

However, the `Go` operating mode differs from the GDA spec when the value is 0: it always prints a plain number, never scientific notation.

Since `Copy` does not copy over the `Context`, the decimal you created with `new` (i.e., `new(decimal.Big).Copy( ... ) ...`) has the `GDA` operating mode. And since you called `Quantize` on both your decimals, both have an exponent of -12 (i.e., a scale of 12).

So, the first time you print it uses the `GDA` operating mode which checks whether the exponent is <= 0 (true, it's -12) and that the exponent plus the length of the coefficient is >= -6 (false, it's -12 + 1 = -11). Since our condition is false, we print a scientific string: `0E-12`.

The second time you print is uses the `Go` operating mode which shortcuts all that when it sees the value of the decimal is 0, and simply returns `0`.

In short, it's because of how string formatting is handled when you have different operating modes. Right now, this only happens when the value is 0.

For more precise string formatting you should check out `Format` method which you can call via `fmt.Sprintf` and friends.

**Comment by tapir** [comment]
Thanks for the detailed explanation.
Is there a reason it doesn't copy context though? I imagine if somebody adheres to a some specific context they would want it to be used when they want a "copy" of the number. This might even cause some unforseeable problems as I wouldn't -for example- expect rounding mode to change when I copy a decimal

**Comment by ericlagergren** [comment]
In general, methods don’t modify the receiver’s (`z`) `Context`. I toyed with the idea of `Copy` being a complete copy, similar to `a := b` (don’t do this, though—the internal `big.Int`s will alias). But that meant there was no way to “set” a decimal without rounding, since `Set` rounds, depending on its context. They were asymmetrical.

I might add something to misc/ or at the very least make the docs more explicit.

**Comment by ericlagergren** [comment]
I should clarify: methods don’t modify the Context *other* than setting conditions.

**Comment by tapir** [comment]
In that case maybe there should be a function NewFromCopy that also copies context? Is that a good solution?

How can I copy the context as well now? is there quick trick?

Thanks

**Comment by ericlagergren** [comment]
WithContext(x.Context).Copy(x) does what you want, but is wordy. But,
that’s what I usually use. You can check out examples if you look at the
math/ subpackage source code.

You could wrap it in a helper function like func copyOf(x *decimal.Big)
*decimal.Big { return ... }

Like I said I might add CopyTotal (or named something similar) to the
music/ subpackafe.

**Comment by tapir** [comment]
Thanks
